Feature/#19 login

mh-algo opened the pull request:
## #️⃣연관된 이슈

> #19 

## 📝작업 내용

> jwt 기반 로그인을 처리하기 위해 Spring Security에 커스텀 인증 필터를 적용했습니다.
> 노션에 로그인 관련 api 명세서도 수정했습니다.
>
> 우선 로그인만 구현된 상태이고 인가는 구현 예정입니다.
> `@AuthenticationPrincipal` 사용하실 땐 `UserPrincipal`를 사용하시면 됩니다.
> jwt로 구현했기에 크기를 최소화하기 위해 principal의 필드를 최소화 했습니다.
> 추가 필드가 필요하실 경우엔 말씀해주시면 반영하겠습니다.


### 스크린샷 (선택)
![로그인 순차 다이어그램](https://github.com/user-attachments/assets/6c8dba24-aabe-4898-ab87-557efa917eee)


## 💬리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
>
> ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요?

mh-algo committed:
feat: #19 SecurityConfig 작성

mh-algo committed:
feat: #19 JwtLoginFilter 구현

mh-algo committed:
chore: #19 R__init_data.sql 수정

- password 수정

mh-algo committed:
chore: #19 MapStruct dependency 추가

mh-algo committed:
feat: #19 AccountService 구현

mh-algo committed:
feat: #19 CustomUserDetailsService 구현

mh-algo committed:
feat: #19 LoginAuthenticationProvider 구현

mh-algo committed:
feat: #19 SecurityConfig 설정 추가

- LoginAuthenticationProvider 추가
- CustomUserDetailsService 추가

mh-algo committed:
chore: #19 패키지 구조 변경

mh-algo committed:
chore: #19 클래스명, 메서드명 변경

- 의도가 명확하게 변경
- JwtLoginFilter.java -> ApiLoginFilter.java
- jwtLoginFilter -> ApiLoginFilter
- getAuthenticationProvider -> getLoginAuthenticationProvider

mh-algo committed:
refactor: #19 UsernameNotFoundException에 cause 추가

mh-algo committed:
chore: #19 JJWT 의존성 추가

mh-algo committed:
feat: #19 JwtUtils 구현

mh-algo committed:
feat: #19 ObjectMapper bean 등록

mh-algo committed:
feat: #19 JwtProvider 구현

- UserPrincipal, LocalAccountDto 사용 필드만 남기고 제거

mh-algo committed:
feat: #19 LoginAuthenticationSuccessHandler 구현

mh-algo committed:
feat: #19 SecurityConfig 설정 추가

- apiLoginFilter에 LoginAuthenticationSuccessHandler 추가

mh-algo committed:
refactor: #19 ApiLoginFilter 생성자 제거

- 사용 안되는 생성자 제거

mh-algo committed:
feat: #19 LoginAuthenticationProvider 비밀번호 패턴 검증 추가

mh-algo committed:
feat: #19 LoginAuthenticationFailureHandler 구현

mh-algo committed:
refactor: #19 filter 등록 방식을 dsl 기반으로 변경

mh-algo committed:
refactor: #19 ErrorCode를 global 패키지에서 security 패키지로 분리

mh-algo committed:
fix: #19 LoginAuthenticationProvider 로그인 검증 순서 변경

- 계정 정지여부 확인 후 비밀번호 비교하는 방식에서 비밀번호 비교 후 계정 정지 상태 확인하는 방식으로 순서 변경

mh-algo committed:
fix: #19 CustomUserDetails의 isEnabled 메서드 수정

- 반환값을 SUSPENDED 상태가 아닌 경우로 수정

mh-algo committed:
chore: #19 dto 클래스 이름 변경

- LocalAccountDto -> LoginAccountInfo

mh-algo committed:
chore: #19 QueryDSL dependency 추가

mh-algo committed:
feat: #19 JPAQueryFactory bean 등록

mh-algo committed:
refactor: #19 로컬 계정 조회시 필요한 컬럼만 조회하도록 변경

- 기존 엔티티 조회 방식에서 QueryDSL을 활용한 객체 조회 방식으로 변경

mh-algo committed:
test: #19 CustomUserDetailsServiceTest 작성

mh-algo committed:
chore: #19 변수명 변경

mh-algo committed:
test: #19 LoginAuthenticationProviderTest 작성

mh-algo committed:
chore: #19 handler 패키지 생성

mh-algo committed:
feat: #19 ApiLoginFilter 예외처리 추가

- json 변환 실패시 400 응답
- post 요청이 아닐 경우 405 응답

mh-algo committed:
test: #19 ApiLoginFilterTest 작성

mh-algo committed:
test: #19 AccountServiceTest 작성

mh-algo committed:
test: #19 LoginAuthenticationSuccessHandlerTest 작성

mh-algo committed:
feat: #19 json property가 올바르지 않은 경우 잘못된 property를 포함하여 응답하도록 변경

mh-algo committed:
test: #19 ApiLoginFilterTest json property 검증 추가

mh-algo committed:
fix: #19 LoginAuthenticationFailureHandler 계정 정지 errorCode 수정

mh-algo committed:
chore: #19 property로 통일

- field와 property 혼용 -> property로 통일

mh-algo committed:
test: #19 LoginAuthenticationFailureHandlerTest 작성

mh-algo committed:
test: #19 JwtProviderTest 작성

mh-algo committed:
feat: #19 ApiLoginFilter 예외 처리

- IOException을 AuthenticationServiceException으로 변환해서 LoginAuthenticationFailureHandler에서 예외 처리하도록 수정

Yuurim98 commented on src/main/java/com/demo/pteam/security/login/ApiLoginFilter.java:
```diff
@@ -0,0 +1,69 @@
+import java.io.IOException;
+
+public class ApiLoginFilter extends AbstractAuthenticationProcessingFilter {
+    private static final ObjectMapper objectMapper = new ObjectMapper();
```
ObjectMapper를 직접 생성하신 이유가 궁금합니다!

mh-algo replied:
`ObjectMapper`의 전역 설정이 필요한 상황은 아니어서 직접 생성해서 사용하는 방식이 더 편할 것 같다고 생각해 이렇게 구현했습니다.
하지만 다시 생각해보니, 객체 지향적 개발 관점에서 봤을 때 생성자로 주입받는게 더 좋을 것 같네요.

Yuurim98 commented on src/main/java/com/demo/pteam/security/login/LoginAuthenticationProvider.java:
```diff
@@ -0,0 +1,57 @@
+        return super.authenticate(authentication);
+    }
+
+    private void validateUsernameFormat(String username, String password) {
```
해당 메서드는 username과 password 둘 다 검증하고 있어서 validateUsernameFormat 보다 다른 이름을 사용하는게 명확할 것 같습니다!

Yuurim98 reviewed (approved):
확인했습니다 고생 많으셨습니다!

seongminQa commented on src/main/java/com/demo/pteam/security/login/ApiLoginFilter.java:
```diff
@@ -0,0 +1,69 @@
+            try {
+                LoginRequest loginRequest = getLoginRequest(request);
+                String username = this.obtainUsername(loginRequest);
+                username = username != null ? username.trim() : "";
```
null일 경우 바로 예외 처리를 하는 것은 어떨까요?

mh-algo replied:
그것도 하나의 방법이라고 생각합니다
저는 단일 책임을 갖도록 구현하는게 좋을 것 같다고 생각해서 filter의 `attemptAuthentication` 메서드에선 request를 처리하는 역할을 하고, provider의 `authentication` 메서드에서 검증을 수행하도록 구현했습니다!

seongminQa reviewed (approved):
고생하셨습니다!!

mh-algo committed:
chore: #19 LoginAuthenticationProvider의 메서드명 변경

- validateUsernameFormat -> validateLoginPattern

mh-algo committed:
refactor: #19 ObjectMapper 생성자 주입 방식으로 변경

mh-algo committed:
test: #19 변경된 코드에 맞춰 테스트 코드 수정

- ObjectMapper 생성자 주입에 맞춰 테스트 코드 수정

mh-algo committed:
feat: #19 json 파싱 실패할 경우 예외 처리 추가

- 파싱 실패시 400 응답

mh-algo committed:
test: #19 로그인 통합테스트 작성

mh-algo committed:
refactor: #19 authenticationManager 설정 방식 변경

- 생성한 AuthenticationManager를 전역 등록한 후 호출해서 필터에 삽입하는 방법 -> configurer의 설정을 통해 필터에 삽입하는 방법으로 변경

mh-algo committed:
refactor: #19 AccountEntity가 SoftDeletableEntity를 상속하도록 수정

0515-Yoonseo-Kim reviewed (approved):
확인했습니다. 고생하셨습니다!

dy1214 reviewed (approved):
확인했습니다!

mh-algo committed:
Merge branch 'dev' into feature/#19-login
